python/lsst/pipe/tasks/configurableActions/_configurableActionStructField.py

```diff
@@ -0,0 +1,330 @@
+
+    def __getattr__(self, attr):
+        if attr in object.__getattribute__(self, '_attrs'):
+            return self._attrs[attr]
+        else:
+            super().__getattribute__(attr)
+
+    def __delattr__(self, name):
+        if name in self._attrs:
+            del self._attrs[name]
+        else:
+            super().__delattr__(name)
+
+    def __iter__(self) -> Generator[ConfigurableAction, None, None]:
+        yield from self._attrs.values()
+
+    def items(self) -> Iterable[Tuple[str, ConfigurableAction]]:
+        yield from self._attrs.items()
+
+
+T = TypeVar("T", bound="ConfigurableActionStructField")
+
+
+class ConfigurableActionStructField(Field):
+    """`ConfigurableActionStructField` is a `~lsst.pex.config.Field` subclass
+    that allows `ConfigurableAction` s to be organized in a
+    `~lsst.pex.config.Config` class in a manor similar to how a
+    `~lsst.pipe.base.Struct` works.
+
+    This class implements a `ConfigurableActionStruct` as an intermediary
+    object to organize the `ConfigurableActions`. See it's documentation for
+    futher information.
+    """
+    # specify StructClass to make this more generic for potential future
+    # inheritance
+    StructClass = ConfigurableActionStruct
+
+    # Explicitly annotate these on the class, they are present in the base
+    # class through injection, so type systems have trouble seeing them.
+    name: str
+    default: Optional[Mapping[str, ConfigurableAction]]
+
+    def __init__(self, doc: str, default: Optional[Mapping[str, ConfigurableAction]] = None,
+                 optional: bool = False,
+                 deprecated=None):
+        source = getStackFrame()
+        self._setup(doc=doc, dtype=self.__class__, default=default, check=None,
+                    optional=optional, source=source, deprecated=deprecated)
+
+    def __set__(self, instance,
+                value: Union[None, Mapping[str, ConfigurableAction], ConfigurableActionStruct],
+                at=None, label='assigment'):
+        if instance._frozen:
+            msg = "Cannot modify a frozen Config. "\
+                  "Attempting to set field to value %s" % value
+            raise FieldValidationError(self, instance, msg)
+
+        if at is None:
+            at = getCallStack()
+
+        if value is None or value == self.default:
+            value = self.StructClass(instance, self, value, at=at, label=label)
+        else:
+            history = instance._history.setdefault(self.name, [])
+            history.append((value, at, label))
+
+        if not isinstance(value, ConfigurableActionStruct):
+            raise FieldValidationError(self, instance,
+                                       "Can only assign things that are subclasses of Configurable Action")
+        instance._storage[self.name] = value
+
+    def __get__(self: T, instance, owner=None, at=None, label="default"
+                ) -> Union[None, T, ConfigurableActionStruct]:
```

Comment:
I'm a bit surprised you've found a tool that's happy with the level of type annotations here (and useful with just that).  I thought they all complained when annotated some arguments/results but not others even within the same method.

Reply:
are we running mypy now in pipe_tasks?

Reply:
No, I am using the language server support in VS code that gives me something like the feedback from just running mypy as I go.

Reply:
@TallJimbo yeah, pylance treats it as an "unknown" type. It does restrict some of the things it can do with downstream analysis, but it does not stop it from running on what it still can. That said I had intended to annotate what I could.